### DISCUSSION TITLE
Enable Caching even after process restart

### DISCUSSION THREAD

**Discussion by prakashpatidarcoindcx** [zilliztech/GPTCache#544] [Q&A] [answered]
I understand below code can use cache stored in database for subsequent same questions based on similarity score.
My question is how to ensure that this caching worked even after process restart . as I have seen that gptcache_answer table is persisted then it should able to use answers stored before process restart.
It will be great help if you could provide modified code for the same .
Thanks in advance:
import os
import numpy as np

from gptcache.adapter import openai
from gptcache import cache
from gptcache.manager import get_data_manager, CacheBase, VectorBase
from gptcache.similarity_evaluation.distance import SearchDistanceEvaluation

d = 8

def mock_embeddings(data, **kwargs):
    return np.random.random((d, )).astype('float32')

def run():
    scalar_stores = [
        CacheBase('sqlite', sql_url='sqlite:///./sqlite.db'),
        CacheBase('postgresql', sql_url='postgresql+psycopg2://postgres:123456@127.0.0.1:5432/postgres'),
        CacheBase('mysql', sql_url='mysql+pymysql://root:123456@127.0.0.1:3306/mysql'),
        CacheBase('mariadb', sql_url='mariadb+pymysql://root:123456@127.0.0.1:3307/mysql'),
        CacheBase('sqlserver', sql_url='ssql+pyodbc://sa:Strongpsw_123@127.0.0.1:1434/msdb?driver=ODBC+Driver+17+for+SQL+Server'),
        CacheBase('oracle', sql_url='oracle+cx_oracle://oracle:123456@127.0.0.1:1521/?service_name=helowin&encoding=UTF-8&nencoding=UTF-8'),
        CacheBase('dynamo'),
    ]

    for scalar_store in scalar_stores:
        if os.path.exists('faiss.index'):
            os.remove('faiss.index')
        vector_base = VectorBase('faiss', dimension=d)
        data_manager = get_data_manager(scalar_store, vector_base)
        cache.init(embedding_func=mock_embeddings,
                   data_manager=data_manager,
                   similarity_evaluation=SearchDistanceEvaluation(),
                   )
        cache.set_openai_key()

        answer = openai.ChatCompletion.create(
            model='gpt-3.5-turbo',
            messages=[
                {'role': 'user', 'content': 'what is chatgpt'}
            ],
        )
        print('answer:', answer)

        answer = openai.ChatCompletion.create(
            model='gpt-3.5-turbo',
            messages=[
                {'role': 'user', 'content': 'what is chatgpt'}
            ],
        )
        print('answer cached:', answer)

if __name__ == '__main__':
    run()

**Comment by SimFG** [comment]
yes, after the process restart, the cache is also effective.

**Comment by prakashpatidarcoindcx** [comment]
Not really SimFG , see screenshot attached .
1.I ran code for prompt what is chatgpt and it has used 136 total tokens 
2.I restarted process and ran again and has used 111 tokens for same question( It should have used cache and 0 tokens but used 111 tokens and without cache)
3.same question without restart is using 0 tokens
any help SimFG to make cache work work after restart for same question asked before restart.

<img width="1596" alt="Screenshot 2023-09-26 at 5 08 58 PM" src="https://github.com/zilliztech/GPTCache/assets/135527669/8b0a7585-20d1-44f5-afc0-db4d1d83cfb8">

**Comment by SimFG** [comment, marked as answer]
![image](https://github.com/zilliztech/GPTCache/assets/21985684/3ceb153b-c640-4479-9ec1-0902d84b1642)
@prakashpatidarcoindcx because the embedding is random and before running the test case, i delete the faiss.index file

**Reply by prakashpatidarcoindcx** [reply]
@SimFG 
I have removed that in my code but it is still not using cache 
<img width="2056" alt="Screenshot 2023-09-27 at 3 02 27 PM" src="https://github.com/zilliztech/GPTCache/assets/135527669/13db2683-2e83-481d-b5de-3343a089f7f5">

**Reply by prakashpatidarcoindcx** [reply]
any help @SimFG 
Can you try the same at your end and confirm?

**Reply by SimFG** [reply]
In fact, I have already told you the reasons, namely
1. the embedding is random
2. delete the faiss.index file

In addition, this example only explains how to set up scalar store. Before using gptcache, you can check the basic documentation, including:

1. [Example usage](https://github.com/zilliztech/GPTCache#example-usage)
2. [Usage, how to use GPTCache better](https://github.com/zilliztech/GPTCache/blob/main/docs/usage.md)
3. [Examples, learn better custom caching](https://github.com/zilliztech/GPTCache/blob/main/examples/README.md)
